usr/src/uts/common/fs/zfs/dsl_dataset.c

```diff
@@ -1971,7 +1972,24 @@ dsl_dataset_space(dsl_dataset_t *ds,
     uint64_t *refdbytesp, uint64_t *availbytesp,
     uint64_t *usedobjsp, uint64_t *availobjsp)
 {
-	*refdbytesp = dsl_dataset_phys(ds)->ds_referenced_bytes;
+	dsl_dir_t *dd = ds->ds_dir;
+	objset_t *os = dd->dd_pool->dp_meta_objset;
+
+	if (dsl_dir_is_zapified(dd) &&
+		zap_contains(os, dd->dd_object, DD_FIELD_COMPRESS_QUOTA) == 0)
+	{
+		uint64_t compress_quota;
+		VERIFY0(zap_lookup(os, dd->dd_object, DD_FIELD_COMPRESS_QUOTA, 8, 1, &compress_quota));
+		if (compress_quota)
+			*refdbytesp = dsl_dataset_phys(ds)->ds_referenced_bytes
+				+ dsl_dataset_phys(ds)->ds_uncompressed_bytes
+				- dsl_dataset_phys(ds)->ds_compressed_bytes;
```

Comment:
I would think that the new quotas would apply to the same concept as the `logicalused` / `logicalreferenced` properties.  As you can see in `dsl_dir_stats()` and `dsl_dataset_stats()`, these are simply `dd_uncompressed_bytes` and `ds_uncompressed_bytes`, respectively.  Is there a reason this seems to be based on a more complicated concept of space accounting?  How would you describe this amount precisely (referenced + uncompressed - compressed)?